### PR TITLE
Dynamically update latex badge menus

### PR DESCRIPTION
**Pull request by jlelong**
This is a follow-up on #3122.

This PR updates the TeX badge when the rootFile changes.

### CODE REVIEW

File: src/main.ts [hunk 1]
```diff
@@ -327,7 +327,8 @@ export class Extension {
         this.texMagician = new TeXMagician(this)
         this.envPair = new EnvPair(this)
         this.section = new Section(this)
-        this.latexCommandTreeView = new LaTeXCommandTreeView()
+        this.latexCommanderProvider = new LaTeXCommanderProvider(this)
```
Comment (tamuratak) [inline src/main.ts:330]: `class Extention` should not have providers. Remove `Extension.latexCommanderProvider`.

File: src/components/commander.ts [hunk 2]
```diff
@@ -15,16 +18,29 @@ export class LaTeXCommandTreeView {
     }
 }
 
-class LaTeXCommander implements vscode.TreeDataProvider<LaTeXCommand> {
-
-    private readonly commands: LaTeXCommand[] = []
+export class LaTeXCommanderProvider implements vscode.TreeDataProvider<LaTeXCommand> {
+    private readonly extension: Extension
+    private readonly _onDidChangeTreeData: vscode.EventEmitter<LaTeXCommand | undefined> = new vscode.EventEmitter<LaTeXCommand | undefined>()
+    readonly onDidChangeTreeData: vscode.Event<LaTeXCommand | undefined>
```
Comment (tamuratak) [inline src/components/commander.ts:24]: Move  `_onDidChangeTreeData` and `onDidChangeTreeData` to `class LaTeXCommandTreeView`

File: src/components/manager.ts [hunk 3]
```diff
@@ -356,8 +356,8 @@ export class Manager {
                 // We need to parse the fls to discover file dependencies when defined by TeX macro
                 // It happens a lot with subfiles, https://tex.stackexchange.com/questions/289450/path-of-figures-in-different-directories-with-subfile-latex
                 await this.parseFlsFile(this.rootFile)
-                this.extension.structureProvider.refresh()
                 this.extension.structureProvider.update()
+                this.extension.latexCommanderProvider.update()
```
Comment (tamuratak) [inline src/components/manager.ts:360]: Define `LaTeXCommandTreeView.update` and call it instead of calling `LatexCommanderProvider.update` directly.

File: src/main.ts [hunk 4]
```diff
@@ -292,6 +291,7 @@ export class Extension {
     readonly texMagician: TeXMagician
     readonly envPair: EnvPair
     readonly section: Section
+    readonly latexCommanderProvider: LaTeXCommanderProvider
```
Comment (tamuratak) [inline src/main.ts:294]: `class Extention` should not have providers. Remove `Extension.latexCommanderProvider`.

**Review by tamuratak — CHANGES_REQUESTED** [review]

File: src/components/commander.ts (hunk 2 above)
Comment (jlelong) [reply src/components/commander.ts:24]: According to https://code.visualstudio.com/api/references/vscode-api#TreeDataProvider%3CT%3E, the event `onDidChangeTreeData` must be defined in the `TreeDataProvider` not in the `TreeView` and the documentation says

>  This will trigger the view to update the changed element/root and its children recursively (if shown).

I could define the `LatexCommanderProvider` directly in the `LaTeXCommandTreeView` and chain the two updates but it would not be coherent with what we do in `structure.ts` and would not really follow the logic of VS Code's API. I think it is better to follow as much as possible the same patterns in all the components.

Even though the `LatexCommanderProvider` is not really needed at the `Extension` level per se, could you expand a bit on the motivations behind 

> `class Extention` should not have providers directly

Comment (tamuratak) [reply src/components/commander.ts:24]: > it would not be coherent with what we do in structure.ts

I don't think what we do in `structure.ts` follows good practice. We had better avoid exposing implementation details. And, `SectionNodeProvider` is an implementation detail. Other components should access it through `StructureTreeView`.

In other words, it is good we have two or more classes, `SectionNodeProvider`, `StructureTreeView`, and others, to implement the structure tree feature. However, we don't have to expose both classes. Hiding implementation details increases the maintainability of the codebase. Splitting a thing into small components is good. Exposing all of them as API is not good.

We can say the same thing with `LaTeXCommandTreeView` and `LaTeXCommanderProvider`. It is good to have multiple classes to implement something. But, we don't have to expose both.

Comment (tamuratak) [reply src/components/commander.ts:24]: > According to https://code.visualstudio.com/api/references/vscode-api#TreeDataProvider%3CT%3E, the event onDidChangeTreeData must be defined in the TreeDataProvider not in the TreeView and the documentation says

Sorry. We don't have to move `onDidChangeTreeData`.

Comment (jlelong) [reply src/components/commander.ts:26]: I have refactored the implementation to stop exposing `LaTeXCommanderProvider`. There was a mix between `LaTeXCommand` and `LaTeXCommander`. Now, we use `LaTeXCommander` everywhere.

**Comment by tamuratak** [comment]
I am not positive about this PR. It increases the maintenance burden. I will never investigate issues related to this feature. Only you have to deal with the related issues. If you are fine with that, I am not particularly against it.

**Comment by tamuratak** [comment]
When recipes dynamically changes, what does `"latex-workshop.latex.recipe.default": "lastUsed"` imply?

**Comment by jlelong** [comment]
> When recipes dynamically changes, what does `"latex-workshop.latex.recipe.default": "lastUsed"` imply?

Good catch! The value of the last used recipe is reset when the root file changes.

**Comment by jlelong** [comment]
> I am not positive about this PR. It increases the maintenance burden. I will never investigate issues related to this feature. Only you have to deal with the related issues. If you are fine with that, I am not particularly against this PR.

I think it is worth giving it a try. If it turns out to create issues we did not think about, I will revert it.
@James-Yu do you have a specific opinion?

File: src/components/manager.ts [hunk 5]
```diff
@@ -358,6 +358,8 @@ export class Manager {
                 await this.parseFlsFile(this.rootFile)
                 this.extension.structureProvider.refresh()
                 this.extension.structureProvider.update()
+                this.extension.builder.reset()
```
Comment (tamuratak) [inline src/components/manager.ts:361]: I don't like resetting `previouslyUsedRecipe` every time rootFile changes. It make `"latex-workshop.latex.recipe.default": "lastUsed"` useless.

**Comment by tamuratak** [comment]
> I think it is worth giving it a try. If it turns out to create issues we did not think about, I will revert it.

I don't think "giving a try and reverting" is an appropriate approach because the recipe is an essential feature. If the structure view doesn't work, users can just ignore it. However, if the recipe doesn't work, it is just a disaster for users. As I said before, tests should be added to this PR at least.

File: src/components/manager.ts (hunk 5 above)
Comment (jlelong) [reply src/components/manager.ts]: Right, I have remove my last commit. I am not a big fan of this `lastUsed` recipe feature. 

https://github.com/James-Yu/LaTeX-Workshop/blob/0ce09bd80e65ffd41658b7b96c917fc007c3feb5/src/components/builder.ts#L429-L446

When `"latex-workshop.latex.recipe.default": "lastUsed"` and the last used recipe name is no more in the recipe list, we peek the first recipe. I think it is a good solution and at least it will not change the current behaviour for single root workspace, which is very important.

**Comment by jlelong** [comment]
> > I think it is worth giving it a try. If it turns out to create issues we did not think about, I will revert it.
> 
> I don't think "giving a try and reverting" is an appropriate approach because the recipe is an essential feature. 

Let me rephrase. If it turns out to create issues requiring too deep changes to be solved, I will revert the PR. I am very confident, that my PR does not change a thing for single rooted workspace, which must represent 99.9% of our users usage otherwise #3162 would have been reported long ago. I have added some tests.

**Comment by jlelong** [comment]
Now, we have to decide if this PR deserves to be merged or not. @tamuratak, @James-Yu what do you think?

**Review by tamuratak — APPROVED** [review]
